jsonrpc/http.go

```diff
@@ -62,6 +75,7 @@ func (h *HTTP) ServeHTTP(writer http.ResponseWriter, req *http.Request) {
 	}
 
 	req.Body = http.MaxBytesReader(writer, req.Body, MaxRequestBodySize)
+	h.requests.Inc()
```

Comment:
Can we expose callbacks for certain events in the "rpc request/response lifecycle"? Otherwise we require dependents to use prometheus.

Reply:
I think I can live with that for now. If we expose callbacks, it will complicate the caller site unnecessarily IMO.

Reply:
Just to make sure we're on the same page: I am proposing something like 

```go
func (s *Server) WithOnRequest(cb func()) *Server {
  s.onRequest = cb
  return s
}
```

This only needs to be set once on the main jsonrpc server, and we don't need to deal with metrics on every protocol bridge (HTTP, websocket, stdio, IPC, etc.). But I'm very open to other ideas.

Reply:
> and we don't need to deal with metrics on every protocol bridge (HTTP, websocket, stdio, IPC, etc.).

I think that is a good metrics tho, to get an idea of which transport protocols are more utilized.

> This only needs to be set once on the main jsonrpc server

Yeap, but that way (if we apply this approach to all packages that collect metrics) we need to move all metric collection code away from the packages and in to `node` package. Since the callbacks and the metrics (counters etc) will be defined on `node` package. That is what I am trying to avoid essentially, to not clutter `node` package.

Reply:
> I think that is a good metrics tho, to get an idea of which transport protocols are more utilized.

Good point.

> we need to move all metric collection code away from the packages and in to node package.

>  That is what I am trying to avoid essentially, to not clutter node package.

Ah ok, that is what I was trying to move towards xD. Move metrics code together into the Juno-specific `node` or `node/metrics` package, instead of sprinkled throughout the codebase. It would be disappointing if virtually every package had to depend on prometheus (which itself pulls in a large number of dependencies).

To avoid excessive use of callbacks, maybe we can compromise with the lightweight [expvar](https://pkg.go.dev/expvar) package? We can combine it with Tailscale's [prometheus exporter](github.com/tailscale/tailscale). It requires us to name metrics a [certain way](https://github.com/tailscale/tailscale/blob/388b12451341494b76976c40467b3785767eddf2/tsweb/varz/varz.go#L34-L36), but I don't think that's too big of an issue.

Reply:
> maybe we can compromise with the lightweight [expvar](https://pkg.go.dev/expvar) package?

We will have to spin up or own implementation for more complex types like `Histogram` or `CounterVec`. I am not sure if it is worth it.

Honestly, prometheus' dependencies are not that bad imo https://github.com/prometheus/client_golang/blob/main/go.mod

Reply:
Prometheus depends on protobuf, which pulls in quite a bit (e.g., see [here](https://github.com/tailscale/tailscale/pull/125#issuecomment-593903976)).

Tailscale [has histograms](https://github.com/tailscale/tailscale/blob/388b12451341494b76976c40467b3785767eddf2/metrics/metrics.go#L79).

Maybe you're right and I'm just over-complicating things. Feel free to go with what you think is best.

Reply:
> I'm just over-complicating things

Nah, you are absolutely right that we should try to minimize our dependencies. But for juno, `prometheus` doesnt really bring many new dependencies, for example we already depend on protobuf.

and at this point, it is so early to try to optimize for re-usability of our packages. We should instead optimize for Juno overall and bring as many features we need as fast as possible.